lib/interceptor/dns.js

```diff
@@ -0,0 +1,125 @@
+import { DecoratorHandler } from '../utils.js'
+import net from 'net'
+import { resolve4 } from 'node:dns/promises'
+
+let fastNow = Date.now()
+setInterval(() => {
+  fastNow = Date.now()
+}, 500)
+
+class Record {
+  address = ''
+  expires = 0
+  errored = 0
+
+  constructor({ address, ttl }) {
+    this.address = address
+    this.expires = fastNow + (ttl ?? 60) * 1e3
+  }
+}
+
+class Handler extends DecoratorHandler {
+  #handler
+  #record
+
+  constructor({ record }, { handler }) {
+    super(handler)
+
+    this.#handler = handler
+    this.#record = record
+  }
+
+  onError(err) {
+    if (
+      [
+        'ECONNRESET',
+        'ECONNREFUSED',
+        'ENOTFOUND',
+        'ENETDOWN',
+        'ENETUNREACH',
+        'EHOSTDOWN',
+        'EHOSTUNREACH',
+        'EHOSTNOTFOUND',
+        'ENODATA',
+        'EPIPE',
+        'UND_ERR_CONNECT_TIMEOUT',
+      ].includes(err.code) ||
+      [503].includes(err.statusCode)
+    ) {
+      this.#record.errored = fastNow
+
+      // TODO (fix): For how long do we "blacklist" the record?
+
+      if (err.code === 'UND_ERR_CONNECT_TIMEOUT') {
+        // We don't expect this address to ever work again...
+        this.#record.expires = Infinity
+      }
+    }
+
+    return super.onError(err)
+  }
+}
+
+export default (interceptorOpts) => (dispatch) => {
+  /** @type {Map<string, Array<Record>>} */
+  const dnsCache = new Map()
+
+  return async (opts, handler) => {
+    if (!opts.dns) {
+      return dispatch(opts, handler)
+    }
+
+    const { protocol, port, hostname, host } = new URL(opts.origin)
+
+    if (net.isIP(hostname) || opts.headers?.host || !port || !protocol) {
```

Comment:
why skip when headers.host or !port?

Reply:
We should always have port and protocol so I guess I should change that to an asser ot default to http:80.

If a host header is set I'm not sure what kind of weird issues it can cause.